### DISCUSSION TITLE
Developing projects with Deno

### DISCUSSION THREAD

**Discussion by zetashift** [aws/aws-pdk#571] [Ideas]
Hi! Happy user of PDK here. It's great with most of the pain points( if not all) actually being JavaScript/TypeScript and its ecosystem.

I don't expect Deno to solve all the problems(it might even bring its own share of problems), but I feel like a `deno` version of: 
`nx-monorepo` and `type-safe-api/typescript`  would help with a few points:
- builtin typescript
- no more config spam in folders because formatting and testing are handled by Deno.
- no more CJS and ESM headaches.

I feel like I need to write more docs about my `.projenrc.ts` because doing anything irregular (e.g. no CJS+React) involves workarounds and just general fiddling that keeps me from writing API's(this isn't projens or pdks fault, this is very much a it is what it is thing). If I can simplify things with Deno I feel like it's atleast worth exploring the space. 
I'm not that well versed in the Nx/projen ecosystem, but here are some interesting links that might be able to get a demo working sooner rather than later:
- https://github.com/nrwl/nx-labs/tree/main/packages/deno
- https://miyauchi.dev/posts/deno-lambda-cdk/
- https://deno.com/blog/npm-on-deno-deploy

I'm not sure what exactly needs to happen to enable it. The simplest start would be to able to generate runtime/handlers `projen` that work well with Deno tooling.

There is also [bun](https://bun.sh/) which is a lot more Node friendly, however I have no experience with that.
Basically: can we simplify the amount of fiddling needed to make a happy TypeScript monorepo that uses `type-safe-api`?

**Comment by cogwirrel** [comment]
This would be really cool - particularly bun! I might just be over-excited because they released their GA youtube video and it sounds very snazzy! 😆 

It could be a good way to reduce build times for type-safe-api projects too as we lose a bit of time waiting for `npx` to run each script, and there are quite a few scripts to run.

**Reply by zetashift** [reply]
Yup I am quite excited to try out bun as well when I have time. But I am entirely unsure how I would go about making a `nx-monorepo` for `bun` instead of `node`. AFAIK they should be compatible. The same goes for the AWS Lambda handlers.
Or is a better first step replacing all `npx` calls with `bunx` calls?

**Reply by cogwirrel** [reply]
Yeah I think there would be plenty of changes dotted around the codebase! I imagine the best place to start would be to add a new `NodePackageManager.BUN` to projen [here](https://github.com/projen/projen/blob/3b0d1067604784d94cb4e3bb6f900ebee846aeab/src/javascript/node-package.ts#L1490) and then following that thread to get the basic `TypeScriptProject` working... I suppose as well as dependencies and `bunx`, you can also use bun as a test runner so might want to swap out the jest stuff from there too...

For the lambdas I guess since AWS Lambda itself uses Node runtime rather than bun, we'd probably need to vend `DockerImageFunction` constructs if we actually want to use bun as the runtime for those :)

**Comment by agdimech** [comment]
Hi - as an FYI we recently added BUN support: https://github.com/aws/aws-pdk/releases/tag/v0.22.34

**Reply by zetashift** [reply]
Awesome!!
